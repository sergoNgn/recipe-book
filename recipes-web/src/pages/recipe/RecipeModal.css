.recipe-modal {
  position: absolute;
  top: 50%;
  left: 50%;
  min-width: 400px;
  transform: translate(-50%, -50%);
  background-color: white;
  border: 0.5px solid #3b3a3a;
  border-radius: 5px;
  box-shadow: 24;
  overflow: auto;
  padding: 1rem;
}
