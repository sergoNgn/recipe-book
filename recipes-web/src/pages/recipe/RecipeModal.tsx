import { FC, useState } from "react";
import { useQueryClient } from "react-query";
import { Modal, Box, TextField, Button } from "@mui/material";
import { notEmpty, maxLength } from "../../validators";
import AlertDialog from "../../components/dialog/AlertDialog";
import { useInput } from "../../hooks/useInput";
import "./RecipeModal.css";
import {
  useRecipeQuery,
  useRecipeSaveQuery,
  useRecipeDeleteQuery,
} from "../../api/queries";

interface RecipeModalProps {
  recipeId: number | undefined;
  onClose: () => void;
  isOpen: boolean;
}

const RecipeModal: FC<RecipeModalProps> = ({ isOpen, onClose, recipeId }) => {
  const queryClient = useQueryClient();
  const [alertDialogOpen, setAlertDialogOpen] = useState(false);

  const { data: recipe } = useRecipeQuery(recipeId);

  const saveRecipeMutation = useRecipeSaveQuery(() => {
    onCloseModal();
    queryClient.invalidateQueries("recipes");
  });

  const deleteRecipeMutation = useRecipeDeleteQuery(() => {
    setAlertDialogOpen(false);
    onCloseModal();
    queryClient.invalidateQueries("recipes");
  });

  const name = useInput(recipe?.name ?? "", [notEmpty()]);
  const shortDescription = useInput(recipe?.shortDescription ?? "", [
    notEmpty(),
    maxLength(150),
  ]);
  const description = useInput(recipe?.description?.data ?? "", [
    notEmpty(),
    maxLength(1000),
  ]);

  const onCloseModal = () => {
    name.reset();
    shortDescription.reset();
    description.reset();
    onClose();
  };

  const canSave = (): boolean => {
    return (
      (name.isDirty || shortDescription.isDirty || description.isDirty) &&
      !(name.isError || shortDescription.isError || description.isError)
    );
  };

  const onSaveRecipe = () => {
    if (canSave()) {
      saveRecipeMutation.mutate({
        ...recipe,
        name: name.value,
        shortDescription: shortDescription.value,
        description: { ...recipe?.description, data: description.value },
      });
    }
  };

  const submitDelete = () => {
    if (recipeId) {
      deleteRecipeMutation.mutate(recipeId);
    }
  };

  return (
    <>
      <AlertDialog
        isOpen={alertDialogOpen}
        handleClose={() => setAlertDialogOpen(false)}
        handleSubmit={submitDelete}
        title="Delete this Recipe?"
        content="Are you shure you want to delete this Recipe?"
      />
      <Modal open={isOpen} onClose={onCloseModal}>
        <Box className="recipe-modal">
          <TextField
            size="small"
            fullWidth
            required
            label="Name"
            margin="normal"
            autoComplete="off"
            error={name.isDirty && name.isError}
            helperText={name.errorText}
            value={name.value}
            onBlur={name.onBlur}
            onChange={name.onChange}
          >
            {name.value}
          </TextField>
          <TextField
            size="small"
            fullWidth
            required
            label="Short Description"
            multiline
            maxRows={4}
            margin="normal"
            autoComplete="off"
            error={shortDescription.isDirty && shortDescription.isError}
            helperText={shortDescription.errorText}
            value={shortDescription.value}
            onBlur={shortDescription.onBlur}
            onChange={shortDescription.onChange}
          >
            {shortDescription.value}
          </TextField>
          <TextField
            fullWidth
            required
            label="Recipe"
            multiline
            minRows={10}
            maxRows={20}
            margin="normal"
            autoComplete="off"
            error={description.isDirty && description.isError}
            helperText={description.errorText}
            value={description.value}
            onBlur={description.onBlur}
            onChange={description.onChange}
          >
            {description.value}
          </TextField>
          <Box className="flex flex-row gap-5 mt-4">
            <Button
              disabled={!canSave()}
              variant="contained"
              className="w-3/12"
              size="small"
              onClick={onSaveRecipe}
            >
              Save
            </Button>
            {recipe?.id && (
              <Button
                onClick={() => setAlertDialogOpen(true)}
                className="w-3/12"
                variant="outlined"
                color="error"
                size="small"
              >
                Delete
              </Button>
            )}
          </Box>
        </Box>
      </Modal>
    </>
  );
};

export default RecipeModal;
