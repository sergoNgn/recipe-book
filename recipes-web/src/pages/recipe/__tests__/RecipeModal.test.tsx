import { act, render, waitFor } from "@testing-library/react";
import RecipeModal from "../RecipeModal";
import { QueryClient, QueryClientProvider } from "react-query";
import { mockRecipe } from "./__mocks__";
import axios from "axios";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

beforeEach(() => {
  jest.clearAllMocks();
});

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const wrapper = ({ children }: any) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

describe("Recipe Modal", () => {
  it("renders recipe modal", async () => {
    mockedAxios.get.mockResolvedValue({ data: mockRecipe });

    const modal = await act(async () =>
      render(<RecipeModal recipeId={1} onClose={() => {}} isOpen={true} />, {
        wrapper,
      })
    );

    expect(modal.baseElement).toMatchSnapshot();
  });
});
