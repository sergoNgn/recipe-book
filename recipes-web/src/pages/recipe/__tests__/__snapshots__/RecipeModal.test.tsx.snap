// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe Modal renders recipe modal 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      class="MuiBox-root css-0"
    />
  </div>
  <div
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="recipe-modal MuiBox-root css-0"
      tabindex="-1"
    >
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r1:"
          id=":r1:-label"
        >
          Name
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            required=""
            type="text"
            value="recipe #1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Name
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r2:"
          id=":r2:-label"
        >
          Short Description
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-z1o969-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r2:"
            required=""
            style="height: 0px; overflow: hidden;"
          >
            short description
          </textarea>
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Short Description
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r3:"
          id=":r3:-label"
        >
          Recipe
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r3:"
            required=""
            style="height: 0px; overflow: hidden;"
          >
            description
          </textarea>
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Recipe
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="flex flex-row gap-5 mt-4 MuiBox-root css-0"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall w-3/12 css-11qr2p8-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Save
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall w-3/12 css-1ir7th6-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Delete
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
